### PR TITLE
Configuration system refactor and new FT.CONFIG command

### PR DESCRIPTION
**Pull request by mnunberg**
Still needs tests

### CODE REVIEW

File: src/config.c [hunk 1]
```diff
@@ -1,125 +1,351 @@
...
-    }
+CONFIG_GETTER(getMinPrefix) {
+  sds ss = sdsempty();
+  return sdscatprintf(ss, "%lld", config->minTermPrefix);
+}
+
+// MAXDOCTABLESIZE
+CONFIG_SETTER(setMaxDocTableSize) {
+  long long size;
+  printf("Hello!!\n");
```
Comment (mnunberg) [inline src/config.c]: TODO: remove this

File: src/config.c [hunk 2]
```diff
@@ -1,125 +1,351 @@
 #include "config.h"
+#include "err.h"
 #include "rmutil/util.h"
 #include "rmutil/strings.h"
 #include <string.h>
 #include <stdlib.h>
+#include <limits.h>
 
-RSConfig RSGlobalConfig = RS_DEFAULT_CONFIG;
+#define RETURN_ERROR(s) return REDISMODULE_ERR;
```
Comment (mnunberg) [inline src/config.c]: When we add richer error reporting, the string will actually be returned to the user somehow

File: src/config.c [hunk 3]
```diff
@@ -1,125 +1,351 @@
...
 
-int ReadConfig(RedisModuleString **argv, int argc, const char **err) {
-  *err = NULL;
-  /* Read the extension we want to load */
-  if (argc > 1 && RMUtil_ArgIndex("EXTLOAD", argv, argc) >= 0) {
-    RMUtil_ParseArgsAfter("EXTLOAD", argv, argc, "c", &RSGlobalConfig.extLoad);
+static int readLongLong(RedisModuleString **argv, size_t argc, size_t *offset, long long *out) {
+  if (*offset == argc) {
```
Comment (MeirShpilraien) [inline src/config.c]: think its better to do '>='

File: src/module.c [hunk 4]
```diff
@@ -1530,6 +1530,44 @@ int AlterIndexCommand(RedisModuleCtx *ctx, RedisModuleString **argv, int argc) {
...
+    return RedisModule_WrongArity(ctx);
+  }
+  const char *action = RedisModule_StringPtrLen(argv[1], NULL);
+  const char *name = RedisModule_StringPtrLen(argv[2], NULL);
+  if (!strcasecmp(action, "GET")) {
+    RSConfig_DumpProto(&RSGlobalConfig, &RSGlobalConfigOptions, name, ctx, 0);
+  } else if (!strcasecmp(action, "HELP")) {
+    RSConfig_DumpProto(&RSGlobalConfig, &RSGlobalConfigOptions, name, ctx, 1);
+  } else if (!strcasecmp(action, "SET")) {
+    size_t offset = 3;  // Might be == argc. SetOption deals with it.
```
Comment (MeirShpilraien) [inline src/module.c]: wouldn't it be easier to deal with 'offset == argc' here? unless there are cases when you do not need value to set configuration but in this case how do you unset it?

File: src/config.c (hunk 3 above)
Comment (mnunberg) [reply src/config.c]: offset should never exceed argc. If it does, I'd like it to crash rather than hide a bug somewhere

File: src/module.c (hunk 4 above)
Comment (mnunberg) [reply src/module.c]: Yeah, that was the idea. Maybe we'd have an unset, or maybe SET without arguments would restore it to the default.

File: src/config.c (hunk 3 above)
Comment (MeirShpilraien) [reply src/config.c]: So lets put an assert on this case

**Pull request merged**